### PR TITLE
Support ray task type checking

### PR DESCRIPTION
**Pull request by pcmoritz**
## Why are these changes needed?

This PR introduces support for typing for ray tasks. For actors, it currently does not do type checking.

Limitation: This supports tasks with up to 10 arguments.

The PR also adds support for some type stubs for the autoscaler.

You can try it out with the following example:

```python
import ray

ray.init()

@ray.remote
def f(a: int) -> str:
  return "a = {}".format(a + 1)

@ray.remote
def g(s: str) -> str:
  return s + " world"

@ray.remote
def h(a: str, b: int) -> str:
  return a

print(f.remote(1))
x = f.remote(1)
print(g.remote(x))

# typechecks but doesn't run                                                                                                                    
print(ray.get(f.remote(x)))

# Does not typecheck:                                                                                                                           
# a = h.remote(1, 1)                                                                                                                            
# b = f.remote("hello")                                                                                                                         
# c = f.remote(1, 1)                                                                                                                            
# d = f.remote(1) + 1
```

You can save it in `example.py` and then run with

```
mypy --strict example.py
```

If you comment in the lines at the bottom of the example, you get an error like

```
example.py:25: error: No overload variant of "remote" of "RemoteFunction" matches argument types "int", "int"
example.py:25: note: Possible overload variant:
example.py:25: note:     def remote(self, arg0: Union[str, ObjectRef], arg1: Union[int, ObjectRef]) -> ObjectRef
example.py:25: note:     <10 more non-matching overloads not shown>
example.py:26: error: No overload variant of "remote" of "RemoteFunction" matches argument type "str"
example.py:26: note: Possible overload variant:
example.py:26: note:     def remote(self, arg0: Union[int, ObjectRef]) -> ObjectRef
example.py:26: note:     <10 more non-matching overloads not shown>
example.py:27: error: No overload variant of "remote" of "RemoteFunction" matches argument types "int", "int"
example.py:27: note: Possible overload variant:
example.py:27: note:     def remote(self, arg0: Union[int, ObjectRef], arg1: Optional[ObjectRef]) -> ObjectRef
example.py:27: note:     <10 more non-matching overloads not shown>
example.py:28: error: Unsupported operand types for + ("ObjectRef" and "int")
Found 4 errors in 1 file (checked 1 source file)
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/worker.pyi [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+@overload
+def remote(function: Callable[[T0, T1, T2, T3, T4, T5, T6, T7], R]) -> RemoteFunction[T0, T1, T2, T3, T4, T5, T6, T7, None, None]: ...
+@overload
+def remote(function: Callable[[T0, T1, T2, T3, T4, T5, T6, T7, T8], R]) -> RemoteFunction[T0, T1, T2, T3, T4, T5, T6, T7, T8, None]: ...
+@overload
+def remote(function: Callable[[T0, T1, T2, T3, T4, T5, T6, T7, T8, T9], R]) -> RemoteFunction[T0, T1, T2, T3, T4, T5, T6, T7, T8, T9]: ...
+# Pass on typing actors for now. The following makes it so no type errors are generated for actors.
+@overload
+def remote(t: type) -> Any: ...
+def remote(function: Callable[..., R]) -> RemoteFunction[T0, T1, T2, T3, T4, T5, T6, T7, T8, T9]: pass
```
Comment (suquark) [inline python/ray/worker.pyi:70]: new line at end of file

File: python/ray/_raylet.pyi [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class ObjectRef(object):
```
Comment (robertnishihara) [inline python/ray/_raylet.pyi:1]: We shouldn't need these `(object)`s, right?

File: python/ray/worker.pyi [hunk 3]
```diff
@@ -0,0 +1,70 @@
+from typing import Any, Callable, Generic, Optional, TypeVar, Union, overload
+
+from ray._raylet import ObjectRef
+
+T0 = TypeVar('T0')
```
Comment (robertnishihara) [inline python/ray/worker.pyi:5]: I guess this file is not being linted, but we use `"` everywhere else.

**Review by robertnishihara — APPROVED** [review]
Please add a test!

**Comment by robertnishihara** [comment]
@pcmoritz We should follow up to make sure that the type checking works well with IDEs. Can you create an issue to track that?

**Pull request merged**